src/Configuration/Programmer/Programmer_cmd.cpp

```diff
@@ -19,57 +19,155 @@ You should have received a copy of the GNU General Public License
+
+  auto formatMemorySize = [](int sizeInBits) -> std::string {
+    const int bitsInKilobit = 1024;
+    const int bitsInMegabit = 1024 * 1024;
+
+    if (sizeInBits >= bitsInMegabit) {
+      int megabits = sizeInBits / bitsInMegabit;
+      return std::to_string(megabits) + "M";
+    } else if (sizeInBits >= bitsInKilobit) {
+      int kilobits = sizeInBits / bitsInKilobit;
+      return std::to_string(kilobits) + "K";
+    } else {
+      return std::to_string(sizeInBits);
     }
+  };
+
+  // openocd path is the same as the executable path
+  std::filesystem::path openOcdExecutablePath = std::filesystem::current_path();
+  std::string openOcdName;
+// if windows
+#ifdef _WIN32
+  openOcdName = "openocd.exe";
+#else
+  openOcdName = "openocd";
+#endif
+  openOcdExecutablePath /= openOcdName;
+
+  int ret = InitLibrary(openOcdExecutablePath.string());
+  if (ret != NoError) {
+    CFG_POST_MSG("Error code = %d. %s", ret, GetErrorMessage(ret).c_str());
+    return ret;
+  }
+  std::vector<Device> devices;
+
+  std::vector<FOEDAG::Cable> cables;
+  ret = GetAvailableCables(cables);
+  if (ret != NoError) {
+    CFG_POST_MSG("Error code = %d. %s", ret, GetErrorMessage(ret).c_str());
+    return ret;
+  } else if (cables.size() == 0) {
+    CFG_POST_MSG("No programming cable found. Make sure it is connected.");
+    return -1;
+  }
+  CFG_POST_MSG("return %d", ret);
+  CFG_POST_MSG("-- GetAvailableCables(cables) --")
+  for (auto& cable : cables) {
+    CFG_POST_MSG("--------------")
+    CFG_POST_MSG("product ID: 0x%04x", cable.productId);
+    CFG_POST_MSG("vendor  ID: 0x%04x", cable.vendorId);
+    CFG_POST_MSG("serial    : %s", cable.serialNumber.c_str());
+    CFG_POST_MSG("desc      : %s", cable.description.c_str());
+    CFG_POST_MSG("bus       : %d", cable.busAddr);
+    CFG_POST_MSG("port      : %d", cable.portAddr);
+    CFG_POST_MSG("device    : %d", cable.deviceAddr);
+    CFG_POST_MSG("speed     : %d", cable.speed);
+    CFG_POST_MSG("transport : %d", cable.transport);
+    CFG_POST_MSG("channel   : %d", cable.channel);
+  }
+
+  CFG_POST_MSG("-- ListDevices(cables, outputMsg) --")
+  ret = ListDevices(cables[0], devices);
+  if (ret != NoError) {
+    CFG_POST_MSG("Error code = %d. %s", ret, GetErrorMessage(ret).c_str());
+    return ret;
+  }
+  if (devices.size() <= 0) {
+    CFG_POST_MSG("No device found. Make sure the board power is ON.");
+    return -1;
+  }
+  CFG_POST_MSG("### ListDevices API testing ###");
+  for (auto& device : devices) {
+    CFG_POST_MSG("-------------------------");
+    CFG_POST_MSG("Device name: %s", device.name.c_str());
+    CFG_POST_MSG("Device index: %d", device.index);
+    CFG_POST_MSG("Device jtagId: 0x%08x", device.tapInfo.idCode);
+    CFG_POST_MSG("Device mask: 0x%08x", device.tapInfo.irMask);
+    CFG_POST_MSG("Device irlength: %d", device.tapInfo.irLen);
+    CFG_POST_MSG("Device flashSize: %s bits",
+                 formatMemorySize(device.flashSize).c_str());
```

Comment:
Any where in the code do the checking of flashsize validity? For example, it must be multiple 8bits, powerof2. Without this checking before calling formatMemorySize() conversion, then the conversion is wrong.

Reply:
The flash size value is read from the return of openOCD. I would think that the validity of the flash size is checked in openOCD level.

The formatMemorySize is a lambda from line 43 to 54. There is no checking in this lambda function as its name implied it just do the formatting.

```auto formatMemorySize = [](int sizeInBits) -> std::string {
    const int bitsInKilobit = 1024;
    const int bitsInMegabit = 1024 * 1024;

    if (sizeInBits >= bitsInMegabit) {
      int megabits = sizeInBits / bitsInMegabit;
      return std::to_string(megabits) + "M";
    } else if (sizeInBits >= bitsInKilobit) {
      int kilobits = sizeInBits / bitsInKilobit;
      return std::to_string(kilobits) + "K";
    } else {
      return std::to_string(sizeInBits);
    }
  };

I can add the checking of flash size value before the formatting happen.

Reply:
The checking should be up-front. It should be done in the API that read devices from openocd (which should be the entry point to any use case), not during the formatting. The validation should be in place for any use case even maybe some use case does not need to display or format.

Reply:
I would like to see, when we rely on other subsystem/module/APIs to get some data, we should always validate the correctness of the data, before we use it.

This should be applicable for general SW development, not only flashsize in this case.

Reply:
Ok. I will add the flash size value check during the ListDevice API entry point.

Reply:
@chungshien I added 
- flash size checking [here](https://github.com/os-fpga/FOEDAG/blob/00f2b400a42fc0af4ca32e8fdc654d636a2ccc4e/src/Configuration/Programmer/Programmer_helper.cpp#L214C1-L216C6)
- the number to unit string conversion [function ](https://github.com/os-fpga/FOEDAG/blob/00f2b400a42fc0af4ca32e8fdc654d636a2ccc4e/src/Configuration/CFGCommon/CFGCommon.cpp#L319C1-L328C2)in cfgcommon 
- added the unit tests for `CFG_convert_number_to_unit_string`